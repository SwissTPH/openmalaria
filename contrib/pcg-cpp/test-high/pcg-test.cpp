/*
 * PCG Random Number Generation for C++
 *
 * Copyright 2014-2017 Melissa O'Neill <oneill@pcg-random.org>,
 *                     and the PCG Project contributors.
 *
 * SPDX-License-Identifier: (Apache-2.0 OR MIT)
 *
 * Licensed under the Apache License, Version 2.0 (provided in
 * LICENSE-APACHE.txt and at http://www.apache.org/licenses/LICENSE-2.0)
 * or under the MIT license (provided in LICENSE-MIT.txt and at
 * http://opensource.org/licenses/MIT), at your option. This file may not
 * be copied, modified, or distributed except according to those terms.
 *
 * Distributed on an "AS IS" BASIS, WITHOUT WARRANTY OF ANY KIND, either
 * express or implied.  See your chosen license for details.
 *
 * For additional information about the PCG random number generation scheme,
 * visit http://www.pcg-random.org/.
 */

/*
 * This file is based on the demo program for the C generation schemes.
 * It shows some basic generation tasks.
 */

#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <cassert>
#include <climits>
#include <iostream>
#include <sstream>
#include <iomanip>
#include <algorithm>
#include <numeric>
#include <random>       // for random_device

#include "pcg_random.hpp"

// This code can be compiled with the preprocessor symbol RNG set to the
// PCG generator you'd like to test.

#ifndef RNG
    #define RNG pcg32
    #define TWO_ARG_INIT 1
#endif

#define STRINGIFY_IMPL(x) #x
#define STRINGIFY(x)      STRINGIFY_IMPL(x)

using namespace std;
using pcg_extras::operator<<;

#if !PCG_EMULATED_128BIT_MATH || !AWKWARD_128BIT_CODE

int main(int argc, char** argv)
{
    // Read command-line options

    int rounds = 5;
    bool nondeterministic_seed = false;

    ++argv;
    --argc;
    if (argc > 0 && strcmp(argv[0], "-r") == 0) {
        nondeterministic_seed = true;
        ++argv;
        --argc;
    }
    if (argc > 0) {
        rounds = atoi(argv[0]);
    }

    /* Many of the generators can be initialized with two arguments; the second
     * one specifies the stream.
     */

#if TWO_ARG_INIT
    RNG rng(42u, 54u);
#else
    RNG rng(42u);
#endif

    if (nondeterministic_seed) {
        // Seed with external entropy from std::random_device (a handy
        // utility provided by pcg_extras).
        rng.seed(pcg_extras::seed_seq_from<std::random_device>());
    }

    constexpr auto bits = sizeof(RNG::result_type) * CHAR_BIT;
    constexpr int how_many_nums = bits <= 8  ? 14
                                : bits <= 16 ? 10
                                :              6;
    constexpr int wrap_nums_at  = bits >  64 ? 2
                                : bits >  32 ? 3
                                :              how_many_nums;

    cout << STRINGIFY(RNG) << ":\n"
    //   << "      -  aka:         " << pcg_extras::printable_typename<RNG>()
    // ^-- we skip this line because the output is long, scary, ugly, and
    //     and varies depending on the platform
         << "      -  result:      " <<  bits << "-bit unsigned int\n"
         << "      -  period:      2^" << RNG::period_pow2();
    if (RNG::streams_pow2() > 0)
         cout << "   (* 2^" << RNG::streams_pow2() << " streams)";
    cout << "\n      -  size:        " << sizeof(RNG) << " bytes\n\n";

    for (int round = 1; round <= rounds; ++round) {
        printf("Round %d:\n", round);

        /* Make some N-bit numbers */
        cout << setw(4) << setfill(' ') << bits << "bit:";
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << endl;

        cout << "  Again:";
        rng.backstep(6);
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << dec << endl;

        /* Toss some coins */
        cout << "  Coins: ";
        for (int i = 0; i < 65; ++i)
            cout << (rng(2) ? "H" : "T");
        cout << endl;

        RNG rng_copy{rng};
        /* Roll some dice */
        printf("  Rolls:");
        for (int i = 0; i < 33; ++i)
            cout << " " << (uint32_t(rng(6)) + 1);
        cout << "\n   -->   rolling dice used "
             << static_cast<size_t>(rng - rng_copy)
             << " random numbers" << endl;

        /* Deal some cards using pcg_extras::shuffle, which follows
         * the algorithm for shuffling that most programmers would expect.
         * (It's unspecified how std::shuffle works.)
         */      
        enum { SUITS = 4, NUMBERS = 13, CARDS = 52 };
        char cards[CARDS];
        iota(begin(cards), end(cards), 0);
        pcg_extras::shuffle(begin(cards), end(cards), rng);

        /* Output the shuffled deck */
        printf("  Cards:");
        static const signed char number[] = {'A', '2', '3', '4', '5', '6', '7',
                                             '8', '9', 'T', 'J', 'Q', 'K'};
        static const signed char suit[] =   {'h', 'c', 'd', 's'};
        int i = 0;
        for (auto card : cards) {
            ++i;
            cout << " " << number[card / SUITS] << suit[card % SUITS];
            if (i % 22 == 0)
                cout << "\n\t";
        }
        
        cout << "\n" << endl;
    }

    // Ensure that the input/output functions work.
    std::stringstream ss;
    ss << rng;
    RNG from_string;
    ss >> from_string;
    assert(from_string == rng);

    return 0;
}

#else //  i.e. PCG_EMULATED_128BIT_MATH && AWKWARD_128BIT_CODE

int main()
{
    // It's not that it *can't* be done, it just requires either C++14-style
    // constexpr or some things not to be declared const.
    cout << "Sorry, " STRINGIFY(RNG) " not supported with emulated 128-bit math"
         << endl;
}

#endif
