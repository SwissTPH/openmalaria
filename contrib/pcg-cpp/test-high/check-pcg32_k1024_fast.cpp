#define RNG pcg32_k1024_fast
#define TWO_ARG_INIT 0

#include "pcg-test.cpp"

