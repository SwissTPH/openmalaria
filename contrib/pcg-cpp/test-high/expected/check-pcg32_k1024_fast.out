pcg32_k1024_fast:
      -  result:      32-bit unsigned int
      -  period:      2^32832
      -  size:        4104 bytes

Round 1:
  32bit: 0xc7221d08 0xff86f771 0x75f34ec3 0xa7a2023a 0x125ae5e7 0xb5ca5bc4
  Again: 0xc7221d08 0xff86f771 0x75f34ec3 0xa7a2023a 0x125ae5e7 0xb5ca5bc4
  Coins: TTTTHTTHTTTTTHHTTHHHHTTTTHHHTTHTHHHHHHTTHTTHTHTHTTTHHTTTHHTHTHTHT
  Rolls: 3 4 2 4 1 5 1 3 4 3 4 3 2 5 2 4 6 1 6 3 3 3 1 3 1 2 3 1 4 1 4 4 6
   -->   rolling dice used 33 random numbers
  Cards: Qs 4h 8h 6d Qd Js Td 2c 3s 8d As Kh Ah Ts 5s Tc 2s 9d 7s 3c Ad 4s
	 Jh 5c Kc 9h 5d 3d Jc Th Kd Qc 4d 2d 6h 9s 5h 6s Ac Ks 7d 7c 8c Qh
	 8s Jd 6c 3h 9c 2h 4c 7h

Round 2:
  32bit: 0x748a6a64 0xa03af0b2 0xb129b59b 0xcf5e4f80 0x4839d150 0xb0ee13b8
  Again: 0x748a6a64 0xa03af0b2 0xb129b59b 0xcf5e4f80 0x4839d150 0xb0ee13b8
  Coins: HTHTHHHHHTTTHHHHTHHHHHTTTHHTHHTHTTTTHTTHTHTTHHTHHTTTTTTHTHTTTTTTH
  Rolls: 5 1 3 6 3 4 5 1 1 1 1 2 3 2 2 4 4 3 6 6 6 6 2 1 2 6 6 2 1 1 3 3 6
   -->   rolling dice used 33 random numbers
  Cards: Qd 2c 7s Ts 7h Ks 3c Kc 2s Js 9h 5c 8d 4s Jd 8c 9c Jc Kd 6c 8s Ac
	 2d 2h 9d Td Qs 7d Tc 6d 9s Ah 4h 5s 3d 3h 7c Qc 8h As 4d 6s Kh Ad
	 Qh Th 6h Jh 5d 5h 3s 4c

Round 3:
  32bit: 0xeed322b4 0x7605071e 0x12f1f90b 0x84d6454e 0x9f738fea 0xa4e3bbb2
  Again: 0xeed322b4 0x7605071e 0x12f1f90b 0x84d6454e 0x9f738fea 0xa4e3bbb2
  Coins: HHTHHHTHHHHTTTHTTHTHHTHTTTHHTTHTTTHTTHTHHTTTTTTTTHHTTHTHHHTHHHHHT
  Rolls: 4 3 3 3 6 2 1 1 4 2 5 5 1 6 2 6 6 2 4 2 2 3 5 6 5 3 3 5 2 5 5 2 5
   -->   rolling dice used 33 random numbers
  Cards: As Kh 4d 2s 3d 3h 4c Qc 8h 6d 7s Qs Kd 9c Ts 6s 7h 6h Ad Qh 9s 4h
	 Jd Ah 6c 3c Jc Ks 5h 8s 2h 7c 3s 8d Js 9h 8c 9d 2d Ac Tc 4s 5c Th
	 Kc Td 7d 2c 5s Jh 5d Qd

Round 4:
  32bit: 0x077dddbd 0x95ae5918 0x4df93a9d 0x35067e2b 0xb3a80eeb 0x621f87bd
  Again: 0x077dddbd 0x95ae5918 0x4df93a9d 0x35067e2b 0xb3a80eeb 0x621f87bd
  Coins: THTHHTTTHHTTHHTTTHHHTHHHTHHHHHHTHTTTHHHHHTHHTHTTHTHHHTTHHTTTTHHTT
  Rolls: 4 2 4 3 2 5 1 2 4 2 3 5 5 6 6 5 4 1 2 4 6 4 3 4 2 1 1 1 4 2 1 2 6
   -->   rolling dice used 33 random numbers
  Cards: 5h Qc 7c 9d 5s 4d 3d 4s Jd Qs Ad 6h 7s 6c 7d Qh 8c Ts 5d 6s Tc 2c
	 3c 8h 6d Qd 3h 4h 8s 9h Th 9c 2d Kc Kh As Js 3s 2s Jh 8d 4c 9s 5c
	 Kd 2h Ah Ac 7h Ks Jc Td

Round 5:
  32bit: 0x4a27d143 0x68c5d423 0xa02ed010 0x8412cad2 0x2356d187 0x9079eac3
  Again: 0x4a27d143 0x68c5d423 0xa02ed010 0x8412cad2 0x2356d187 0x9079eac3
  Coins: HHTHHTTTHHHTTHHTHTHHTHTTHHTHHTHHHTHHTTTHTTHTHHTTHHTTHHHHTTTTHTHTH
  Rolls: 1 6 4 1 1 4 6 1 1 5 3 2 4 5 2 6 5 1 4 2 5 4 3 6 2 4 2 5 4 5 6 3 4
   -->   rolling dice used 33 random numbers
  Cards: 5c 4d Tc Kh Qc 6h 9h 7d 7s Th 4s 5s 6s Ks 6c 3d 7h 7c 2d Ad 5h Kd
	 9c 8d 4h Qh 9s 2c 4c 5d 8h 6d Kc 3c Ts 3s As Qd Ah Jd Qs 2s 2h Jh
	 Td Jc 8c Js 9d Ac 8s 3h

